api/api-scms-v4/src/main/java/com/thoughtworks/go/apiv4/scms/SCMControllerV4.java

```diff
@@ -0,0 +1,231 @@
+        super(ApiVersion.v4);
+        this.apiAuthenticationHelper = apiAuthenticationHelper;
+        this.pluggableScmService = pluggableScmService;
+        this.entityHashingService = entityHashingService;
+        this.goConfigService = goConfigService;
+    }
+
+    @Override
+    public String controllerBasePath() {
+        return Routes.SCM.BASE;
+    }
+
+    @Override
+    public void setupRoutes() {
+        path(controllerBasePath(), () -> {
+            before("", mimeType, this::setContentType);
+            before("/*", mimeType, this::setContentType);
+            before("", mimeType, this::verifyContentType);
+            before("/*", mimeType, this::verifyContentType);
+
+            before("", this.mimeType, this.apiAuthenticationHelper::checkAdminUserOrGroupAdminUserAnd403);
+            before("/*", this.mimeType, this.apiAuthenticationHelper::checkAdminUserOrGroupAdminUserAnd403);
+
+            get("", mimeType, this::index);
+            post("", mimeType, this::create);
+            get(Routes.SCM.ID, mimeType, this::show);
+            put(Routes.SCM.ID, mimeType, this::update);
+            delete(Routes.SCM.ID, mimeType, this::destroy);
+            get(Routes.SCM.USAGES, mimeType, this::getUsages);
+        });
+    }
+
+    public String index(Request request, Response response) throws IOException {
+        SCMs scms = pluggableScmService.listAllScms();
+
+        String etag = entityHashingService.hashForEntity(scms);
+        if (fresh(request, etag)) {
+            return notModified(response);
+        }
+        setEtagHeader(response, etag);
+        return writerForTopLevelObject(request, response, writer -> SCMsRepresenter.toJSON(writer, scms));
+    }
+
+    public String show(Request request, Response response) throws IOException {
+        String materialName = request.params(MATERIAL_NAME);
+
+        SCM scm = fetchEntityFromConfig(materialName);
+
+        String etag = entityHashingService.hashForEntity(scm);
+        if (fresh(request, etag)) {
+            return notModified(response);
+        }
+        setEtagHeader(response, etag);
+        return writerForTopLevelObject(request, response, writer -> SCMRepresenter.toJSON(writer, scm));
+    }
+
+    public String create(Request request, Response response) {
+        SCM scmFromRequest = buildEntityFromRequestBody(request, false);
+
+        scmFromRequest.ensureIdExists();
+
+        haltIfEntityWithSameNameExists(scmFromRequest);
+        haltIfEntityWithSameIDExists(scmFromRequest);
+
+        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+        pluggableScmService.createPluggableScmMaterial(currentUsername(), scmFromRequest, result);
+
+        return handleCreateOrUpdateResponse(request, response, scmFromRequest, result);
+    }
+
+    public String update(Request request, Response response) {
```

Comment:
How are we handling updates to SCM's from ConfigRepo?

Reply:
Previously it sends 404 saying that the said scm id was not found.
Updated it to return 422 with the message saying it is defined in the config repo